### PR TITLE
chore: upgrade yarn to v4 and other dependencies

### PR DESCRIPTION
**Pull request by abnud11**
---
name: Upgrade dependencies of root project
about: it's about upgrading deps of the project
title: "Upgrade dependencies of root project"
labels: ""
assignees: "abnud1"
---

## Description

**Problem**
This PR upgrades yarn from v1 to v4 using corepack, it also upgrades other deps to the latest version, especially major updates like `react-testing-library`, Typescript from 5.4 to 5.5, and others.

This PR also upgrades CodeQL Github action to use v3.
**Changes**

I updated `package.json` along with `yarn.lock`, `codeql-analysis.yml` file, I also added `@types/react` and `@types/react-dom` and enabled `skipLibCheck: true` in tsconfig in order to make `yarn type-check` script successful

## Contribution checklist
- [x] I have followed the [contributing guidelines](https://github.com/Hacker0x01/react-datepicker/blob/main/CONTRIBUTING.md).
- [x] I have added sufficient test coverage for my changes.
- [x] I have formatted my code with Prettier and checked for linting issues with ESLint for code readability.

### CODE REVIEW

**Comment by abnud11** [comment]
After upgrading `react-testing-library` to v16, some tests failed.

For some reason, `fireEvent.keyDown` is not firing sometimes, the resolution to those cases are documented here:
https://testing-library.com/docs/guide-events#keydown

In short, you need to focus on the element to dispatch keydown, then dispatch keydown on `document.activeElement || document.body`, while this should be done in all the cases where `fireEvent.keyDown` is used, I used it only in the failing cases, not sure why it succeeds sometimes and fails other times.

**Comment by abnud11** [comment]
Can we approve the workflows and merge this PR please ?

File: .yarn/install-state.gz
Comment (martijnrusschen) [inline .yarn/install-state.gz:1]: what's this?

File: docs-site/.yarn/install-state.gz
Comment (martijnrusschen) [inline docs-site/.yarn/install-state.gz:1]: Same

File: docs-site/public/index.html
Comment (martijnrusschen) [inline docs-site/public/index.html:1]: why is is this related?

File: examples/hello-world/.yarn/install-state.gz
Comment (martijnrusschen) [inline examples/hello-world/.yarn/install-state.gz:1]: Uninteded it seems.

Comment (martijnrusschen) [reply examples/hello-world/.yarn/install-state.gz:1]: perhaps these files should be added to gitignore.

File: package.json [hunk 1]
```diff
@@ -22,67 +36,71 @@
...
+    "@testing-library/react": "^16.0.0",
     "@types/eslint": "^8.56.10",
     "@types/jest": "^29.5.12",
     "@types/node": "22",
-    "@typescript-eslint/eslint-plugin": "^7.9.0",
-    "@typescript-eslint/parser": "^7.9.0",
-    "axe-core": "^4.4.1",
-    "babel-jest": "^29.6.4",
+    "@types/react": "^18.3.4",
+    "@types/react-dom": "^18.3.0",
```
Comment (martijnrusschen) [inline package.json:59]: These 2 are new? Why?

File: package.json [hunk 2]
```diff
@@ -22,67 +36,71 @@
...
-    "jest-environment-jsdom": "^29.6.4",
-    "lint-staged": "^15.0.1",
+    "jest-environment-jsdom": "^29.7.0",
+    "lint-staged": "^15.2.9",
     "lodash": "^4.17.21",
-    "prettier": "^3.2.5",
-    "react": "^18.2.0",
-    "react-dom": "^18.2.0",
-    "rollup": "^4.13.0",
+    "postcss": "^8.4.41",
```
Comment (martijnrusschen) [inline package.json:79]: why is this needed?

File: package.json (hunk 1 above)
Comment (abnud11) [reply package.json:59]: `yarn type-check` which is part of Github test workflow fails without these 2 packages, honestly don't know why it didn't error before.

File: .husky/pre-commit
Comment (martijnrusschen) [inline .husky/pre-commit:1]: Why is this removed?

Comment (abnud11) [reply .husky/pre-commit:1]: This is part of husky 9 upgrade, while husky was on version 9 before this PR, this change should have been done when migrating from 8 to 9, see this:
https://github.com/typicode/husky/releases/tag/v9.0.1

In particular, read `How To Migrate`

**Comment by martijnrusschen** [comment]
Ok thanks let's resolve the last conflict and merge this.

**Comment by abnud11** [comment]
@martijnrusschen conflicts fixed.

**Comment by martijnrusschen** [comment]
Something is wrong with Eslint. I can't get it to work anymore. It seems to consistently fail now, and as you can see in this PR, the linter didn't run but did succeed.

**Comment by martijnrusschen** [comment]
Same seems to happen with sass-lint

**Comment by abnud11** [comment]
@martijnrusschen it works on my machine correctly, what exactly doesn't work on your machine? did you run `yarn install`?

**Comment by abnud11** [comment]
also, the linter ran as part of `update (20.x)` workflow

**Comment by abnud11** [comment]
And `sass-lint` works too

**Comment by martijnrusschen** [comment]
Example: https://github.com/Hacker0x01/react-datepicker/actions/runs/10652202374/job/29525711646

**Comment by abnud11** [comment]
@martijnrusschen that job fails at `yarn test:ci` not `yarn lint`, `yarn lint` actually passes and I don't think the test failure is related to this PR.

**Comment by martijnrusschen** [comment]
but if you compare it to an older run, let's say this one: https://github.com/Hacker0x01/react-datepicker/actions/runs/10460719357/job/28967495012. you will see more output for the lint job. It should run 2 linters, and it doesn't seem to run normally.

**Comment by abnud11** [comment]
That's because it was using yarn v1, with yarn v4 it doesn't show what commands it runs, only their output if there is any output, `eslint` and `stylelint` doesn't show any output unless there is an error.

I confirmed on my local machine that `eslint` catches errors by introducing errors in Typescript files, I also ensured that `stylelint` catches errors by introducing errors to sass code too.
Without introduced errors, they don't output anything since there are no errors.

@martijnrusschen
